shards/modules/physics/shapes.cpp

```diff
@@ -0,0 +1,281 @@
+    std::visit(
+        [&](auto &&arg) {
+          using T = std::decay_t<decltype(arg)>;
+          if constexpr (std::is_same_v<T, gfx::MeshDrawable::Ptr>) {
+            this->addDrawable(arg, linalg::identity);
+          } else if constexpr (std::is_same_v<T, gfx::MeshTreeDrawable::Ptr>) {
+            this->addDrawable(arg);
+          } else {
+            throw SHException("Invalid input type");
+          }
+        },
+        drawable.drawable);
+  }
+
+  void addDrawable(const gfx::MeshTreeDrawable::Ptr &drawable) {
+    pmr::SharedTempAllocator tempAllocator;
+    gfx::TransformUpdaterCollector collector(tempAllocator.getAllocator());
+    collector.ignoreRootTransform = true;
+    collector.collector = [&](const gfx::DrawablePtr &drawable, const float4x4 &worldTransform) {
+      if (gfx::MeshDrawable::Ptr ptr = std::dynamic_pointer_cast<gfx::MeshDrawable>(drawable)) {
+        this->addDrawable(ptr, worldTransform);
+      }
+    };
+    collector.updateNoModify(*drawable);
+  }
+
+  void addDrawable(const gfx::MeshDrawable::Ptr &drawable, const float4x4 &worldTransform) {
+    auto &mesh = drawable->mesh;
+    if (!mesh) {
+      SPDLOG_LOGGER_WARN(logger, "Ignoring drawable with no mesh");
+      return;
+    }
+    addMesh(mesh, worldTransform);
+  }
+
+  void addMesh(const gfx::MeshPtr &mesh, const float4x4 &worldTransform) {
+    auto &srcFormat = mesh->getFormat();
+
+    auto &vertexData = mesh->getVertexData();
+
+    std::optional<size_t> positionIndex;
+    std::optional<size_t> positionOffset;
+
+    size_t srcStride = srcFormat.getVertexSize();
+    size_t offset{};
+    for (size_t i = 0; i < srcFormat.vertexAttributes.size(); i++) {
+      auto &attrib = srcFormat.vertexAttributes[i];
+      if (attrib.name == "position") {
+        positionIndex = i;
+        positionOffset = offset;
+        break;
+      }
+      offset += attrib.numComponents * getStorageTypeSize(attrib.type);
+    }
+
+    // Validate required attributes
+    {
+      if (!positionIndex.has_value()) {
+        throw std::runtime_error("Mesh does not have a position attribute");
+      }
+      auto positionAttrib = srcFormat.vertexAttributes[positionIndex.value()];
+      if (positionAttrib.type != gfx::StorageType::Float32 && positionAttrib.numComponents != 3) {
+        throw std::runtime_error("position attribute must be float3");
+      }
+    }
+
+    {
+      const uint8_t *basePtr = vertexData.data() + *positionOffset;
+      for (size_t i = 0; i < mesh->getNumVertices(); i++) {
+        auto &pt = *(float3 *)(basePtr + srcStride * i);
+        float3 transformedP = linalg::mul(worldTransform, float4(pt, 1.0)).xyz();
+        points.push_back(toJPHVec3(transformedP));
+      }
+    }
+  }
+};
+
+struct MeshHullShapeShard {
+  static SHTypesInfo inputTypes() { return MeshSrcTypes::Types; }
+  static SHTypesInfo outputTypes() { return SHShape::Type; }
+  static SHOptionalString help() { return SHCCSTR(""); }
```

Comment:
Please don't do this, we will think documentation exists and ignore it in the priority matrix.

Reply:
Just document it briefly

Reply:
Fixed